Target agnostic interface for 8-bit matrix multiply for wasm

abhi-agg opened the pull request:
This PR add the necessary interface for 8-bit matrix multiply function (`uint8 * int8`) for wasm

Fixes: https://github.com/browsermt/bergamot-translator/issues/213

I request reviewers to please verify whether the the function signatures are correct for the intended use case.

_Open issues that will be resolved in this PR_ incrementally:
1. Which one of `Int8::PrepareB` and `Int8::PrepareBTransposed` functions should be added to make the interface complete while still keeping it target agnostic. I will add them once we reach a decision.
     _**EDIT**_: **We agreed to add Int8::PrepareB.**

2. Do we need to add a fused Multiply function (a multiply function with PrepareA fused inside)?
    _**EDIT**_: **We agreed that it can be added later based on if this function can provide performance gains.**

3. Should we support `int8 * int8` multiplication as well? If yes, then I will add `Int8::Multiply` and the corresponding `Int8::PrepareA` function to support this use case.
     _**EDIT**_: **We agreed that we don't need to add separate functions for this. The same Multiply function can deal with both.**


### Checklist

- [ ] I have tested the code manually
- [ ] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

yurydelendik commented:
Such data organization is not really practical and will take a performance hit on bounds checks: you have to check now if `QuantizedBuffer&` structure in bounds, then deference and read `value` pointer, as well as deference `Index*` cols, and perform bounds checks there. We want to keep data for intrinsic in registers as much as possible; any reference/pointer will force data to be read from/written into memory.

XapaJIaMnu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,54 @@
+typedef struct QuantizedBuffer {
+    int8_t* value;
+    float scale;
+    int8_t zero_point;
```
@kpu  we want `float` zero_point, right?

kpu replied:
So that would be the float value that int8_t 0 corresponds to?  Then it may as well be:
int8_t *value;
float min_range /* -128 maps to this */, max_range; /* 127 maps to this */
Which is more flexible.  

I had been thinking 0 is always somewhere in the quantized range and the int8_t value tells us which one maps to 0.  Because so many values are near 0 it's usually important to have an exact 0 quantization point though it could be expressed in the above form.

abhi-agg replied:
So the `zero_point` stays as `int8_t`?

abhi-agg replied:
@kpu What is the final call on this one?

XapaJIaMnu replied:
@kpu I was thinking of, taking matrix A, subtracting a float value X, so that everything else is zero centered and then quantising, But in practise we just want to not care about outliers.

abhi-agg replied:
@kpu A gentle ping 👍

kpu replied:
I am ok with @XapaJIaMnu 's more general proposal that zero_point be float.

abhi-agg replied:
Great. Changing it to `float` then 👍

XapaJIaMnu replied:
@abhi-agg the zero point is not a `float` yet.

abhi-agg replied:
Done.

XapaJIaMnu reviewed (commented):
As talked in the other document, we need at least one more version of the the PrepareB functions. Furthermore it is not really practical to have a struct, as structs may exhibit different alignments on different hardware. Normally what is done is to stick extra metadata at the end of the long array (eg at the end of the `int8_t*` we have 1 float for the scale of A, one for the scale of B, and 2 zero_points, one corresponding to A and one to the B matrix).

XapaJIaMnu commented:
As for point 2., we will write some experimental kernels where you take `float` matrix A and `int8_t` matrix B and see if they are competitive in terms of speed compared to the two pass approach.

abhi-agg commented:
@yurydelendik @XapaJIaMnu If I understood correctly, removing QuantizedBuffer struct and adding the 3 fields of this struct (`int8_t* value, float scale, int8_t zero_point`) as function parameters will resolve the concern. Right?

XapaJIaMnu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,54 @@
+
+/* Prepares bias for the Multiply routine that performs unsigned * signed multiplication.
+ */
+void Int8ShiftPrepareBias(const QuantizedBuffer& B_input, const float* bias_input, QuantizedBuffer& output, Index width, Index B_cols);
```
PrepareBias outputs a float, not an int8_t.

abhi-agg replied:
As per the document (`Multiply(Quantized A, Quantized B, Quantized Bias)`, the bias needed for Multiply function is quantized. Was it a typo? Just double checking here.

abhi-agg replied:
@XapaJIaMnu So, a prepared B is taken (which is int8_t*) as input which is de-quantized to float and then an offset (float bias) is applied to it to give the final bias (as float). Right?

XapaJIaMnu replied:
No. Let's say B is of dimensions MxN. Then the bias would be of dimension 1xN. You perform a multiplication with a column matrix 1xM which contains just ones and then the result is multiplied by this: https://github.com/marian-nmt/marian-dev/blob/78e856e2d8a205587114071c8cfd9c2a0c7c7ebe/src/tensors/cpu/intgemm_interface.h#L394
And added to the bias.

XapaJIaMnu commented:
> @yurydelendik @XapaJIaMnu If I understood correctly, removing `QuantizedBuffer` struct and adding its 3 fields (`int8_t* value, float scale, int8_t zero_point`) as function parameters will resolve the concern. Right?

The zero point should be float, but yeah. Waiting for feedback from @kpu

abhi-agg commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,77 @@
+void Int8SelectColumnsB(const int8_t* B_input,
+                        Index B_rows,
+                        Index B_cols,             // We need this for bound checking for B buffer?
+                        const Index* cols_begin,  // Shouldn't cols_begin be just Index and not Index*?
+                        const Index* cols_end,    // Shouldn't cols_end be just Index and not Index*?
```
@XapaJIaMnu @kpu Could you please help me in understanding what SelectColumnsB function does?

Does it select columns of B from `cols_begin` to `cols_end` index with the condition that `0 <= cols_begin, cols_end < B_cols`

if yes, then shouldn't `cols_begin` and `cols_end` arguments be just plain `Index` instead of `Index*`?

XapaJIaMnu replied:
SelectColumsnB selects subset of columns from a matrix. So what you need is an array of indecis to select (let's say [1, 25, 8723, 9837]) this is why you need ptrs. In practise this is usually the `begin()` and `end()` of a vector: https://github.com/marian-nmt/marian-dev/blob/78e856e2d8a205587114071c8cfd9c2a0c7c7ebe/src/tensors/cpu/intgemm_interface.h#L213

abhi-agg replied:
Okay. So the array of indices can also be represented as `(const Index* cols_begin, Index num_cols)`. In this case, `cols_end` will be equal to `cols_begin + num_cols - 1`

Thanks for the explanation.

XapaJIaMnu replied:
Let's keep it the same way it's implemented in the function to avoid unnecessary complexity.

yurydelendik replied:
>  the array of indices can also be represented as (const Index* cols_begin, Index num_cols). In this case, cols_end will be equal to cols_begin + num_cols - 1

Actually it is just cols_end = cols_begin + num_cols.

There is no difference for underlined implementation (since second pointer is easy to calculate). In any case, you will have to know exact size of `Index` (not `unsigned int` of course) for layout of `cols_XXX` array.

kpu replied:
@yurydelendik is correct we are half-open interval C++ people

abhi-agg replied:
Okay. Then I will go with `(const Index* cols_begin, Index num_cols)` where array half-open interval
[cols_begin, cols_begin+num_cols)

yurydelendik replied:
Keep in mind, on spidermonkey side, as safety property we need:
- To validate `cols_end`: check if it is in wasm memory boundary and `(cols_end - cols_begin)` is positive and multiple of `Index`;
- or, to validate `num_cols`: check if it is positive and `cols_begin + num_cols * sizeof (Index)` is in wasm memory boundary.

Both require about the same about of ops.

abhi-agg commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,77 @@
+
+#include <cstdint>
+
+using Index = unsigned int;
```
I will change `Index` to a fixed-width data type `unit32_t`.

XapaJIaMnu replied:
We use `unsigned int` https://github.com/kpu/intgemm/blob/master/intgemm/types.h#L56 . We may need to negotiate this change before you change it unliterary

abhi-agg replied:
@kpu @XapaJIaMnu Could there be any issue if I change it to `uint32_t`? It would be great if we can adopt a fixed-width data type in these APIs.

kpu replied:
64-bit support is a good thing to have in the general library but I would be ok with an option.

XapaJIaMnu replied:
I *think* not, but we should be consistent when using those datatypes.

yurydelendik replied:
This interface is for FFI purpose only, original intgemm code mostly shall stay unaffected. My recommendation is to define different type e.g. `IndexForFFI` define it as e.g. `uint32_t`, use in this interface instead of `Index`, and implementation assert this assumption in the library?

abhi-agg replied:
@kpu @XapaJIaMnu  Should I go ahead with `uint32_t` then? We can use assert for `sizeof(Index) == sizeof(IndexForFFI)` as @yurydelendik suggested.

XapaJIaMnu replied:
Is there are problem if we use `uint64_t` since this is what *should* be used to address 64 bit arrays. Otherwise we may run into issue with (very) big matrices.

yurydelendik replied:
to be flexible `size_t` has to be use, which reflect addressable pointer size

yurydelendik replied:
Defining `Index` as `uint64_t` will cripple performance on 32-bit platforms. Defining as `unsigned int` is on 64-bit and 32-bit is mostly, but not guaranteed, is 32 bit. `uint32_t` just makes it official.

abhi-agg replied:
`size_t` is `4-bytes` for wasm platform and I had encountered a bug in marian because `size_t` and `uint64_t` were being used interchangeably.

However, `Index` parameter is being used to refer to the indices of an array and from that logic it should be `size_t`.

What is the final call on this? Can the `Index` ever be bigger than the max value of `uint32_t`?

yurydelendik replied:
My recommendation to set it to uint32_t in this header. (In the implementation of the function you can just assert `sizeof(Index) == sizeof(uint32_t)`. Only 16-bit platforms will be affected by explicitly making `uint32_t` here.

abhi-agg replied:
Changing `Index` to `uint32_t` as
1. It is set to `unsigned int` in intgemm's existing interface
2. The number of model parameters are hardly going to touch such a high number

Please let me know if anyone has a concern. I am making the change.

abhi-agg commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,77 @@
+void Int8ShiftPrepareBias(const int8_t* B_input,
+                        float scale,
+                        int8_t zero_point,
+                        const float* bias_input,
```
@kpu @XapaJIaMnu What would be the dimension of `bias_input` parameter? Would it be same as of B (i.e B_rows * B_cols).

XapaJIaMnu replied:
Biases a 1xB_cols

abhi-agg replied:
@XapaJIaMnu As per this comment of yours https://github.com/browsermt/marian-dev/pull/49#discussion_r683295414 (specifically https://github.com/marian-nmt/marian-dev/blob/78e856e2d8a205587114071c8cfd9c2a0c7c7ebe/src/tensors/cpu/intgemm_interface.h#L394), I see both `quant_mult_a` and `quant_mult_b` being used to prepare bias.

However, I am only providing `quant_mult_b` via this function. Should I also provide scale for A as a function parameter?

XapaJIaMnu replied:
So, the actual routine takes only one `float`. We usually compute that `float` in the caller.

abhi-agg replied:
Okay. Then I can leave it as it is.

abhi-agg commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,77 @@
+
+/* Main interface for integer matrix multiplication (C = A * B) for wasm.
+ *
+ * A is typically activations whose rows should be a multiple of 1 (i.e. no restriction) and
+ * columns should be a multiple of 64.
+ * 
+ * B is typically fixed model parameters whose rows should be a multiple of 64 and columns
+ * should be a multiple of 8.
```
A question for documentation purpose. Would A and B matrices also be in row-major format?

XapaJIaMnu replied:
We assume everything is in a Row Major format.

abhi-agg replied:
Great. I will document it then 👍

abhi-agg commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,77 @@
+                        const float* bias_input,
+                        Index B_rows,
+                        Index B_cols,
+                        float* output);
```
The dimension of the `output` matrix would then be 1 x B_cols. Right?

XapaJIaMnu replied:
The input bias is the same shape as the output bias, so yes.

XapaJIaMnu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,159 @@
+                int8_t zero_point,
+                Index B_rows,
+                Index B_cols,
+                int8_t* output);
```
Add an argument here: `bool transpose=false`

abhi-agg replied:
@XapaJIaMnu Just to make sure that I understood the function of `bool transpose` argument correctly, could you please confirm following for me:

Assuming shape of the final output matrix `output` is `(B_rows, B_cols)`

1. When `transpose=true`  => shape of input matrix `B_input` is expected to be `(B_cols, B_rows)`
2. When `transpose=false` => shape of input matrix `B_input` is expected to be `(B_rows, B_cols)` (i.e same as `output` matrix)

XapaJIaMnu replied:
Yes, I think this is the correct behaviour.

Our intgemm interface differs, but we should just fix that.

abhi-agg replied:
Great. Adding the parameter and the documentation around it.

abhi-agg replied:
Done. Resolving this as I pushed a commit for this.

XapaJIaMnu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,159 @@
+
+
+/**
+ * A multiply routine to perform multiplication of 2 matrices.
```
To avoid confusing others who may want to use this, we should point out that A_input is actually `uint8_t`, not `int8_t`.

abhi-agg replied:
@XapaJIaMnu Wouldn't specifying this expose the architecture specific information in the interface? IIUC (and please correct me if I am wrong here):

1. The implementations of this interface for architectures that have a native `uint8 * int8` instruction (e.g. Intel and very recent Arm architecture) will do the following:
    1. `PrepareA` function will prepare A such that all of its entries are non-negative (e.g. by adding 127)
    2. `PrepareBias` will prepare a valid bias for the Multiply function
    3. `Multiply` function will perform multiplication internally using the corresponding native instruction (uint8 * int8)

2. The implementation of this interface for architectures that have a native `int8 * int8` instruction (e.g. Arm architecture) will do the following:
    1. `PrepareA` function will prepare A **_but it will not do anything to make all of its entries non-negative_**.
    2. `PrepareBias` will prepare **_an empty bias_** (as bias comes into picture only for supporting `uint8 * int8` case).
    3. `Multiply` function will perform multiplication internally using the corresponding native instruction (int8 * int8)

i.e. Most of the functions of this interface will have an architecture specific implementations but the user should not care about those details.

I can add in the documentation (wherever missing) that this interface implements things internally in architecture dependent manner. So the correct use of Multiply function requires using the corresponding `Prepare* `functions to prepare `A`, `B` and `bias`.

cc @kpu

XapaJIaMnu replied:
So, we always want to use `int8*int8` if this is available, as it simplifies everything massively, however this is not available on x86 (for any widely available hardware at least).

I think that we can add a function such as `isSignedTimesSigned()` that the client can use to determine the correct codepath. On top of that, we can also just support an identity `PrepareBias()` that does nothing.

abhi-agg replied:
> So, we always want to use `int8*int8` if this is available, as it simplifies everything massively, however this is not available on x86 (for any widely available hardware at least).
> 

This is fine.

What I wanted to say is that we don't need to specify any implementation specific details in the interface (like you were saying here https://github.com/browsermt/marian-dev/pull/49#discussion_r685318208). User should be oblivious to them.

> I think that we can add a function such as `isSignedTimesSigned()` that the client can use to determine the correct codepath. On top of that, we can also just support an identity `PrepareBias()` that does nothing.

I believe we don't need this. AFAIR, spidermonkey will detect this on its own and call the right implementation. Right @yurydelendik  @eqrion ?

yurydelendik replied:
I don't understand the question. The spidermonkey will not control any logic. If will just bridge the interface between wasm code and native (intrinsic) code. Whatever is implemented on either side will be your choice.

abhi-agg replied:
> It will just bridge the interface between wasm code and native (intrinsic) code. Whatever is implemented on either side will be your choice.

As per my understanding, spidermonkey will determine which target architecture the code is running on and, based on this information, will hook the corresponding implementation of the matrix multiply. Please correct me if I am wrong here.

yurydelendik replied:
I see intgemm is already doing cpuid. Unless you trust spidermonkey more to do that for the library, intgemm can keep doing it. Spidermonkey at this moment only detect if general platform is supported, e.g. ARM64 vs Intel, and enables intrinsics. Would you like spidermonkey to fo cpuid for you?

abhi-agg replied:
> I think that we can add a function such as `isSignedTimesSigned()` that the client can use to determine the correct codepath. On top of that, we can also just support an identity `PrepareBias()` that does nothing.

@XapaJIaMnu Since we are providing one implementation of matrix multiply for each architecture (Intel SSSE3, AVX2, AVX512, Armv8.0, Armv8.4/8.5), wouldn't a cpu detection logic suffice to determine which implementation should be used under the hood? By `isSignedTimesSigned()`, you meant a cpu detection logic. Right?

XapaJIaMnu replied:
Currently if you're doing intel, you're always doing UnsignedTimesSigned, whereas if you're doing ARM you're doing SignedTimesSigned. It would be wasteful for the one implementing the multiplication to always force the presence of a bias so that the multiplication result is correct (Especially given that several big machine translation providers are moving away from biases for some of the matrices). We should allow users to choose a more efficient codepath when possible.

abhi-agg replied:
For this, I believe the solution is pretty straightforward. I can add one more Multiply function that doesn't take any bias argument.
In all, there will exist 2 Multiply functions:

1. `MultiplyAndAddBias(A, B, Bias, output)`
    1. It will do `unsigned * signed` multiplication by using the trick `A * B + Bias`
    2. It will require a valid implementation of `PrepareBias`
2. `Multiply(A, B, output)`
    1. It will do `signed * signed` multiplication by just `A * B`
    2. It doesn't require `PrepareBias`

However, the 2nd `Multiply` function can be added later as well when you provide the Arm implementation.

abhi-agg replied:
I changed the name of the existing `Multiply` function to `MultiplyAndAddBias` and added documentation that this function can be used for supporting matrix multiplication on architectures that don't have a native instruction for signed*signed but have a native instruction for unsigned*signed

abhi-agg replied:
@XapaJIaMnu I probably don't understand what the issue is here. If the concern is that how will we support both `signed * signed` and `unsigned * signed` then I see 2 options:

1. We support both cases using single Multiply function. This means:
    1. For architectures having only `unsigned * signed` native instruction (e.g. Intel)
        1. `PrepareA` will prepare A so that every value is non-negative and
        2. `PrepareBias` prepares a valid bias for the `Multiply` function
    2. For architectures having `signed * signed` native instruction (e.g. ARM)
        1. `PrepareA` will not try to make every value non-negative
        2. `PrepareBias` prepares an empty bias (or identity bias as you were saying)
2. We support both cases using 2 different Multiply functions:
    1. For `signed * signed`, something like `Multiply(A, B)`
        1. This requires implementing only PrepareA and PrepareB 
    2. For `unsigned * signed`, something like `MultiplyAndAddBias(A, B, Bias)`
        1. This requires implementing PrepareA, PrepareB and PrepareBias

During last plenary meeting we decided that Option 1 is a satisfactory approach. Yes, in this option, I agree with you that the architectures that support `signed * signed` will be forced to provide implementation of `PrepareBias` even though bias is not needed for multiply.

Do you have some other concern that I am missing here?

kpu replied:
I'm debating a system like this:

B and bias are always carried together in an opaque blob.  There's a `void PackBAndBias(int8_t *blob, const float *B, const float *bias, Index B_rows, Index B_cols, float scaling)` that allows the bias term to be `nullptr` for no bias.  The blob would be variable size (it might create a bias term or not; it might be padded), so there's another call to tell the user how much memory it needs `size_t SizeBAndBias(Index B_rows, Index B_cols, bool have_bias)`.  

A is packed in an opaque blob.  There's a `void PackA(int8_t *blob, const float *A, Index A_rows, Index B_cols, float scaling)` and maybe an analogous `size_t SizeA(Index A_rows, Index A_cols)` though we don't need one now because it's just A_rows * A_cols.  

One Multiply function gets PackBAndBias's blob and PackA's blob.  It already knows if there is bias because PackB put it in the opaque blob.  On Intel there always will be bias.  If Nick wants to do ARM without bias, he can.  

There's also a flavor of Multiply that goes directly from float for A.

abhi-agg replied:
@kpu @XapaJIaMnu Please let me know your thoughts on the last comment 👍

kpu replied:
I think we should go ahead without the fanciness of not having bias and variable sizing.  Then only if it's justified by the speed it can be extended.

abhi-agg replied:
Okay. I will leave everything as it is then 👍

kpu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,193 @@
+                  float scale,
+                  int8_t zero_point,
+                  bool is_input_transposed,
+                  Index output_rows,
```
If it applies to both the output and input, why is it called output?

kpu replied:
Also why is the user specifying the shape of an opaque output representation that is neither row nor column major?

XapaJIaMnu replied:
@abhi-agg we should have the comment, pointing out that if there is a transposition happening, the rows/cols is reversed. This is typical in BLAS style interfaces, so users will be familiar with this peculiarity. Refer to this naming:

https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L339

abhi-agg replied:
@XapaJIaMnu Please check out the latest version. I have tried to be descriptive there. If that still doesn't feel enough, please feel free to comment and I will address it 👍

abhi-agg replied:
Resolving it for now.

kpu commented:
I think it would be a helpful exercise for @abhi-agg to write short implementations of the proposed functions in terms of the existing functions for several reasons:
1. You'll have a working implementation.  
2. The cost of binding it to implementation will disincentivize @abhi-agg from unnecessary changes that appear in the present API.  

I worry the relatively cheap cost of speccing out a new API (when one already exists) will result in a tax on having to change the implementation.  

For example, it is not clear why the function argument order has changed.  Or what value making the user specify the output shape of a matrix is (as opposed to the input shape), when it's actually going to an opaque representation that is neither row major or column major.

kpu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,193 @@
+                  float scale,
+                  int8_t zero_point,
+                  Index output_rows,
+                  Index output_cols,
```
Why is this called output_cols while the other one is called cols_B?

abhi-agg replied:
I will change it to be consistent everywhere.

kpu commented:
While it's not explicitly stated, intgemm's interface always talks to the user in row major (unless the function name has Transposed) and that's the meaning of A_rows, inner dimension, and B_cols.  These refer to A and B in row major format.  The opaque representation (B) is private to intgemm.  These sizes refer to the input.  There is a convention.  

Here the convention has been flipped to output sizes; is there a reason for this?  For prepareB this is undefined.  From an API usability perspective, I feel like callers know the shape of their input best.

abhi-agg commented:
@kpu

> For example, it is not clear why the function argument order has changed

I can make the order same as of intgemm interface 👍 

> While it's not explicitly stated, intgemm's interface always talks to the user in row major (unless the function name has Transposed) and that's the meaning of A_rows, inner dimension, and B_cols. These refer to A and B in row major format. The opaque representation (B) is private to intgemm. These sizes refer to the input. There is a convention if not a stated one.
>

Taking an example just to make sure that I understood it correctly. Do you mean that:
1. The argument `rows` of [PrepareA(const float *input, int8_t *output, float quant_mult, Index rows, Index cols)](https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L188) and `A_rows` of [Multiply(const int8_t *A, const int8_t *B, Index A_rows, Index width, Index B_cols, Callback callback)](https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L219) are same and represent the no. of rows of input matrix A (`input`)?
2. The argument `rows` of [(*PrepareB)(const float *input, int8_t *output, float quant_mult, Index rows, Index cols)](https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L201), `rows` of [(*SelectColumnsB)(const int8_t *input, int8_t *output, Index rows, const Index *cols_begin, const Index *cols_end)](https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L215) and `width` of [Multiply(const int8_t *A, const int8_t *B, Index A_rows, Index width, Index B_cols, Callback callback)](https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L219) are all the same and represent the no. of rows of input matrix B?

Additionally, could you please explain what you mean by `opaque` representation? This would be super helpful.

XapaJIaMnu commented:
@abhi-agg opaque representation refers to the fact, that after `PrepareB` the output `B` matrix is neither rowM, nor ColumnMajor, but an opaque binary representation. You can iterate it and make sense of it.

andrenatal commented:
Can we move this faster and expedite it? Our WebAssembly team is blocked by the decision that needs to be taken here, and if we keep delaying it, we might lose their support and let this slip through the cracks. 

What's still missing to have this landed, @kpu @XapaJIaMnu @abhi-agg?

Thanks

abhi-agg commented:
> @abhi-agg opaque representation refers to the fact, that after `PrepareB` the output `B` matrix is neither rowM, nor ColumnMajor, but an opaque binary representation. You can iterate it and make sense of it.

@XapaJIaMnu Thanks a lot. This clarifies a lot. Few questions:
1. Is the output of `PrepareA` and `PrepareBias` also opaque?
2. The result of the multiply function (C) is in row-major format. Right?

XapaJIaMnu commented:
@abhi-agg 
1) `PrepareA` and `PrepareBias` are both row Major format, but it is difficult to explain to the user what exactly `prepareBias` does, as I illustrated it to you beforehand.
2) `C` is RowMajor.

Yes for prepareA and Multiply,
For PrepareB and Multiply, it's a bit more complicated. If B is supposed to be transposed while it is being prepare (PrepareBTransposed), it corresponds width corresponds to the rows of the transposed matrix (or the cols of the original one). See the signature of this function.

https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L339

What you have written holds for the default case where B doesn't need to be transposed before the operation:

https://github.com/kpu/intgemm/blob/master/intgemm/intgemm.h#L328

abhi-agg commented:
@XapaJIaMnu Thanks a lot. Kenneth's and your answers have cleared a lot of unstated conventions of intgemm interface.

I have changed the documentation to reflect all of it. However, it is possible that there are few things which I may have misunderstood. So, it would be great if you can give it a final look and point out anything that seems wrong 👍 

Additionally:
1. I have added a separate function for taking the transposed B input. I believe the API stays clean that way (instead of a boolean flag to pass a transposed and non-transposed B matrix in the same function).
2. I haven't changed the order of the function arguments to make it consistent with intgemm as intgemm doesn't seem to follow a consistent pattern (there is inconsistency in the order of input and output arguments across functions that use a callback and the ones that don't). The current interface follows a simple pattern: output is always the last argument and this makes the interface pretty consistent.
3. I have changed the documentation to use rows and cols of the input matrices and not of the intermediate results (i.e. prepared A, prepared B and prepared Bias).
4. While reviewing, please look out for the places where I shouldn't have mentioned about the Shape of the matrix.

@andrenatal I believe one final review from Nik should make it possible to land this.

andrenatal commented:
thanks @abhi-agg and @XapaJIaMnu

XapaJIaMnu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,246 @@
+ * @param[in]   zero_point     The zero point (for quantization)
+ * @param[in]   rows_A         No. of rows of Input matrix A. No restriction on its size.
+ * @param[in]   width          No. of columns of Input matrix A. It should be a multiple of 64.
+ * @param[out]  output         An array representing the prepared A matrix.
```
Should  mention that on some architectures (x86) the output would be `uint8_t` (Which is achieved by adding 127 to every item in A after quantising), in case someone decides to quantise `A` without using this routine.  

We don't need to change the data type, as we can do ptr casting underneath, but we should probably mention that different architectures will have different underlying behaviour, which will be opaque to the user if they strictly use only the provided functions.

abhi-agg replied:
> in case someone decides to quantise A without using this routine.

@XapaJIaMnu  I have mentioned in the documentation that the precondition to use the Multiply function is that A, B and Bias should be prepared using Prepare* functions of this interface otherwise there is no guarantee of the correctness of the result. Should I still add the information regarding prepared A to be `uint8` for x86?

XapaJIaMnu replied:
I believe that additional documentation never hurts. We can have an additional implementation details at the bottom of the document, that may help devs understand better what we are doing. After all our goal is not to produce an obfuscated binary interface. I can write this at a later point, but I think this is an ok interface.

What are the next steps? We need to write implementations of those functions that map to the actual intgemm calls?

abhi-agg replied:
> We can have an additional implementation details at the bottom of the document, that may help devs understand better what we are doing. After all our goal is not to produce an obfuscated binary interface. I can write this at a later point, but I think this is an ok interface.

okay.




> We need to write implementations of those functions that map to the actual intgemm calls?

Yes, this is the next step.

I am merging the PR then 👍

XapaJIaMnu commented on src/tensors/cpu/wasm_intgemm_interface.h:
```diff
@@ -0,0 +1,246 @@
+ *
+ * Please note that this interface might have architecture specific implementation.
+ *
+ * @param[in]   input_A_prepared       An array representing the prepared A matrix.
```
Again mention that this input is architecture dependent (`uint8_t` on x86).

XapaJIaMnu reviewed (approved):
I think it looks good, just some more warnings about the data type of a prepared A.
